packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/JSONNotificationContentBuilder.java

```diff
@@ -0,0 +1,129 @@
+package expo.modules.notifications.notifications;
+
+import android.net.Uri;
+import android.util.Log;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import androidx.annotation.Nullable;
+import expo.modules.notifications.notifications.enums.NotificationPriority;
+import expo.modules.notifications.notifications.model.NotificationContent;
+
+public class JSONNotificationContentBuilder extends NotificationContent.Builder {
+  private static final String TITLE_KEY = "title";
+  private static final String TEXT_KEY = "message";
+  private static final String SUBTITLE_KEY = "subtitle";
+  private static final String SOUND_KEY = "sound";
+  private static final String BODY_KEY = "body";
+  private static final String VIBRATE_KEY = "vibrate";
+  private static final String PRIORITY_KEY = "priority";
+  private static final String BADGE_KEY = "badge";
+
+  public JSONNotificationContentBuilder() {
+  }
+
+  public NotificationContent.Builder setPayload(JSONObject payload) {
+    this.setTitle(getTitle(payload))
+        .setSubtitle(getSubtitle(payload))
+        .setText(getText(payload))
+        .setBody(getBody(payload))
+        .setPriority(getPriority(payload))
+        .setBadgeCount(getBadgeCount(payload));
+    if (shouldPlayDefaultSound(payload)) {
+      useDefaultSound();
+    } else {
+      setSound(getSound(payload));
+    }
+
+    if (shouldUseDefaultVibrationPattern(payload)) {
+      useDefaultVibrationPattern();
+    } else {
+      setVibrationPattern(getVibrationPattern(payload));
+    }
+    return this;
+  }
+
+  protected String getTitle(JSONObject payload) {
+    try {
+      return payload.getString(TITLE_KEY);
+    } catch (JSONException e) {
+      return null;
+    }
+  }
+
+  protected String getText(JSONObject payload) {
+    try {
+      return payload.getString(TEXT_KEY);
+    } catch (JSONException e) {
+      return null;
+    }
+  }
+
+  protected String getSubtitle(JSONObject payload) {
+    try {
+      return payload.getString(SUBTITLE_KEY);
+    } catch (JSONException e) {
+      return null;
+    }
+  }
+
+  protected Number getBadgeCount(JSONObject payload) {
+    try {
+      return payload.has(BADGE_KEY) ? payload.getInt(BADGE_KEY) : null;
+    } catch (JSONException e) {
+      return null;
+    }
+  }
+
+  protected boolean shouldPlayDefaultSound(JSONObject payload) {
+    return !payload.optBoolean(SOUND_KEY, true);
```

Comment:
Oh yeah, I think the non-clearness comes from how `optBoolean` is implemented. Let me cite the implementation:
```java
    public boolean optBoolean(String name, boolean fallback) {
        Object object = opt(name);
        Boolean result = JSON.toBoolean(object);
        return result != null ? result : fallback;
    }
```
where `JSON.toBoolean` is implemented like:
```java
      static Boolean toBoolean(Object value) {
        if (value instanceof Boolean) {
            return (Boolean) value;
        } else if (value instanceof String) {
            String stringValue = (String) value;
            if ("true".equalsIgnoreCase(stringValue)) {
                return true;
            } else if ("false".equalsIgnoreCase(stringValue)) {
                return false;
            }
        }
        return null;
    }
```

so, to sum up, if:
- `"sound"` is `true` or `"true"`: `optBoolean(SOUND_KEY, true)` will return `true`
- `"sound"` is `false` or `"false"` `optBoolean(SOUND_KEY, true)` will return `false`
- `"sound"` is anything else: `optBoolean` will fallback to the fallback value, which is `true`

What I wished to support was:
- `true | "true" | false | "false"` to decide whether to play a default sound
- `null` not to play any sound
- `string` to play a sound from the value interpreted as `Uri`
- fallback to `true` otherwise.

Indeed so, there is a bug in here, which should have been fixed just now.

Reply:
Gotcha. But with the `!` in the current logic, if `"sound"` is true or undefined then `shouldPlayDefaultSound` will return false, whereas if `"sound"` is false then `shouldPlayDefaultSound` will return true, no? This seems backwards -- did you mean to remove the `!`?

Reply:
Of course you were right, sorry!

Current implementation is:

```java
  protected boolean shouldPlayDefaultSound(JSONObject payload) {
    try {
      // value is an explicit boolean
      return payload.getBoolean(SOUND_KEY);
    } catch (JSONException e) {
      // do nothing
    }

    // do not play a default sound only if the value is a valid Uri
    return getSound(payload) != null;
  }

  protected Uri getSound(JSONObject payload) {
    try {
      payload.getBoolean(SOUND_KEY);
      // value is an explicit boolean
      return null;
    } catch (JSONException e) {
      // it's not a boolean, let's handle it as a string
    }

    try {
      if (payload.optString(SOUND_KEY) != null) {
        return Uri.parse(payload.getString(SOUND_KEY));
      }
    } catch (JSONException | NullPointerException e) {
      // do nothing
    }
    return null;
  }
```

so:
- if `sound` is an URI string, we don't want to play default sound unless the URI is invalid (Android handles this for us)
- if `sound` is set to `true` or `"true"` or `false` or `"false"`, we interpret it as a boolean
- otherwise we fallback to true (if `sound` is not set or it is set to some non-falsy value—we already handled all the falsy values [`null` values aren't delievered])

Reply:
(sorry, missed this notification somehow! 😞 )

Awesome, thanks for writing it out like this -- I think it's clearer even though it's longer & more verbose.

I still think there is one tiny issue -- should the last line of `shouldPlayDefaultSound` be `return getSound(payload) == null;` instead?

This is exactly the kind of thing it would be great to have native unit tests for. 😅 Let's not worry about that for the moment, though!